app/models/ems_refresh/save_inventory.rb

```diff
@@ -52,85 +52,91 @@ def save_vms_inventory(ems, hashes, target = nil)
-            found = found_dups.select { |v| v.ems_ref == h[:ems_ref] }
-            found = found_dups if found.empty?
-          end
-        end
-        found = found.first
 
-        if found.nil?
-          _log.info("#{log_header} Creating Vm [#{h[:name]}] location: [#{h[:location]}] storage id: [#{h[:storage_id]}] uid_ems: [#{h[:uid_ems]}] ems_ref: [#{h[:ems_ref]}]")
+    ActiveRecord::Base.transaction do
+      hashes.each do |h|
+        # Backup keys that cannot be written directly to the database
+        key_backup = backup_keys(h, remove_keys)
+
+        h[:ems_id]                 = ems.id
+        h[:host_id]                = key_backup.fetch_path(:host, :id) || key_backup.fetch_path(:host).try(:id)
+        h[:ems_cluster_id]         = key_backup.fetch_path(:ems_cluster, :id) || key_backup.fetch_path(:ems_cluster).try(:id)
+        h[:storage_id]             = key_backup.fetch_path(:storage, :id)
+        h[:storage_profile_id]     = key_backup.fetch_path(:storage_profile, :id)
+        h[:flavor_id]              = key_backup.fetch_path(:flavor, :id)
+        h[:availability_zone_id]   = key_backup.fetch_path(:availability_zone, :id)
+        h[:cloud_network_id]       = key_backup.fetch_path(:cloud_network, :id)
+        h[:cloud_subnet_id]        = key_backup.fetch_path(:cloud_subnet, :id)
+        h[:cloud_tenant_id]        = key_backup.fetch_path(:cloud_tenant, :id)
+        h[:cloud_tenants]          = key_backup.fetch_path(:cloud_tenants).compact.map { |x| x[:_object] } if key_backup.fetch_path(:cloud_tenants, 0, :_object)
+        h[:orchestration_stack_id] = key_backup.fetch_path(:orchestration_stack, :id)
+        begin
+          raise MiqException::MiqIncompleteData if h[:invalid]
+
+          # Find the Vm in the database with the current uid_ems.  In the event
+          #   of duplicates, try to determine which one is correct.
+          found = indexed_vms[h[:uid_ems]] || []
+
+          if found.length > 1 || (found.length == 1 && found.first.ems_id)
+            found_dups = found
+            found = found_dups.select { |v| v.ems_id == h[:ems_id] && (v.ems_ref.nil? || v.ems_ref == h[:ems_ref]) }
+            if found.empty?
+              found_dups = found_dups.select { |v| v.ems_id.nil? }
+              found = found_dups.select { |v| v.ems_ref == h[:ems_ref] }
+              found = found_dups if found.empty?
+            end
+          end
+          found = found.first
 
-          # Handle the off chance that we are adding an "unknown" Vm to the db
-          h[:location] = "unknown" if h[:location].blank?
+          if found.nil?
+            _log.info("#{log_header} Creating Vm [#{h[:name]}] location: [#{h[:location]}] storage id: [#{h[:storage_id]}] uid_ems: [#{h[:uid_ems]}] ems_ref: [#{h[:ems_ref]}]")
 
-          # build a type-specific vm or template
-          found = ems.vms_and_templates.build(h)
-        else
-          vms.delete(found)
```

Comment:
I don't think you can delete this line.  Its purpose, I *think* it to ensure that the same vm isn't referenced twice.  That is, if two vms have the same uid_ems, then on the first pass of the loop it might grab the first one, and we want to ensure that on the second pass of the loop it grabs the second one.  I think all that has to happen is to remove it from the new `indexed_vms`